ops/ccs-ops-misc/synthetic-data/scripts/synthea-automation/s3_utilities.py

```diff
@@ -0,0 +1,358 @@
+from botocore.exceptions import ClientError, WaiterError
+from pathlib import Path
+
+boto_config = Config(region_name="us-east-1")
+s3_client = boto3.client('s3', config=boto_config)
+
+# wait config for checking presence of manifest.xml in /Done folder
+# check very 30 seconds
+s3_wait_delay = 30
+# willing to wait for up to 3 days:
+#   1440 mins / day * 3 days == 4320
+#   multiply times 2 since we'll try every 30 secs
+s3_wait_max_retries = 8640
+
+# generic FQN for persisting end_state.properties
+end_state_props_file = "end_state/end_state.properties"
+
+# Mitre S3 bucket
+mitre_synthea_bucket = "bfd-synthea"
+
+# list of proprietary Mitre mapping files that will be downloaded;
+# needed to generate synthetic data.
+code_map_files = [
+    'betos_code_map.json',
+    'condition_code_map.json',
+    'dme_code_map.json',
+    'drg_code_map.json',
+    'hcpcs_code_map.json',
+    'medication_code_map.json',
+    'snf_pdpm_code_map.json',
+    'snf_pps_code_map.json',
+    'snf_rev_cntr_code_map.json',
+    'external_codes.csv'
+    ]
+
+# Mitre shell script files that need to be downloaded; used to generated synthetic data.
+code_script_files = [
+    'national_bfd.sh',
+    'national_bfd_v2.sh'
+    ]
+
+# Function to download proprietary Mitre mapping files from an S3 bucket; the files
+# are necessary to generate synthetic data. The mapping filenames are listed in an
+# array of filenames so this function loops over names in the code_map_files array.
+#
+# Param: target_dir : unix filesystem directory where downloaded S3 file(s) are written to.
+# Raises a python exception if failure to download file.
+def download_synthea_files(target_dir):
+    for fn in code_map_files:
+        output_fn = target_dir + fn
+        try:
+            s3_client.download_file(mitre_synthea_bucket, fn, output_fn)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                print(f"The object does not exist: {fn}")
+            else:
+                raise
+
+# Function to download proprietary Mitre shell script files from an S3 bucket; the scripts
+# drive the synthea generation process. The scripts are listed in an array of filenames
+# so this function loops over names in the code_script_files array. Since the files will
+# be run as shell script(s), unix file permissions are set to enable them to be executed.
+#
+# Param: target_dir : unix filesystem directory where downloaded S3 file(s) are written to.
+# Raises a python exception if failure to download file.
+def download_synthea_scripts(target_dir):
+    for fn in code_script_files:
+        output_fn = target_dir + fn
+        try:
+            s3_client.download_file(mitre_synthea_bucket, fn, output_fn)
+            os.chmod(os.fspath(output_fn), 0o744)
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == "404":
+                print(f"The object does not exist: {fn}")
```

Comment:
Will this error fail the script by propagating anything back up? If this fails we basically can't continue, so I'd imagine we should just blow everything up if this happens. (Same for some of the other errors in this file)

Reply:
yea, it would have; I basically cleaned up all exception handling to absolutely (re-) throw exceptions if anything is off.